### PR TITLE
Release serialization resources (fixes #4774)

### LINKED ISSUE #4774: Runtime Error SYSTEM_NO_ROLL in export package
Runtime rrror occurs from function export package.

```
Category               Resource Shortage
Runtime Errors         SYSTEM_NO_ROLL
Date and Time          14.05.2021 11:58:44



 Short Text
     The memory request for 33865852 bytes could not be complied with.



 What happened?
     Each transaction requires some main memory space to process application
     data. If the operating system does not have any more memory, the
     transaction is terminated.



 What can you do?
     Try to find out (using specific data selection for example) whether the
     transaction will run with less main memory.

     If the problem is due to a temporary bottleneck, run the transaction
     again.

     If the error persists, ask your system administrator to check the
     following profile parameters:

     o  ztta/roll_area            (1.000.000 - 15.000.000) classic roll area
     per user and internal session o  ztta/roll_extension       (10.000.000
     - 500.000.000) available memory/user in extended memory
     EM  o  ztta/max_memreq_MB maximum size of an individual memory request
     (in MB) o  abap/heap_area_total         (100.000.000 - 1.500.000.000)
     available memory (malloc) for all users of an application server. If
     several background processes are running on one server, temporary
     bottlenecks can occur. Of course, the amount of memory (in bytes) must
     also be available on the machine (main memory or file system swap).
     Caution:
     The operating system must be set up so that there is also enough memory
     for each process. Usually, the maximum address space is too small.
     Ask your hardware manufacturer or your competence center about this.

     o  abap/heap_area_dia:          (10.000.000 - 1.000.000.000)
     restriction of the memory that can be allocated to the heap with malloc
     per dialog process.

     Parameter for background processes:

     o  abap/heap_area_nondia:       (10.000.000 - 1.000.000.000)
     restriction of the memory that can be allocated to the heap with malloc
     per background process.

    Other memory-relevant parameters are:

    o  em/initial_size_MB:          (35 - 1200)
    Extended memory area from which all users of an application server can
    meet their memory needs.

    o  abap/shared_objects_size_MB:
    Size of the shared memory area that can be used to save shared objects
    area instances. This memory area is global for all users of the
    application server.
    Note which actions and entries caused the error to occur.

    Consult your SAP administrator.

    Using transaction ST22 for ABAP dump analysis, you can view, manage,
    and retain termination messages for longer periods.



Error analysis
    More main memory area was requested.
    The total available space has already been assigned however. Possible
    causes:
    - A large number of (large) internal tables
    - A large number of (large) programs active
    - Deep nesting of subroutines with large amounts of local data



How to correct the error
    Perform an analysis to find out whether this request makes sense or
    whether a program error has occurred. Pay particular attention to the
    internal table entries listed below.
    The amount of memory (in bytes) used at termination time was:

    Roll area...................... 0
    Extended memory (EM) 2004191024
    Assigned memory (HEAP) 2050793808
    Short area.....................
    Paging area.................... 106496
    Maximum address space.......... " "
    If the error occurs in a non-modified SAP program, you might be able to
    find a solution in the SAP Notes system. If you have access to the SAP
    Notes system, check there first using the following keywords:

    "SYSTEM_NO_ROLL"
    "CL_ABAP_ZIP===================CP" or CL_ABAP_ZIP===================CM006
    "SAVE"
    If you cannot solve the problem yourself, please send the following
    information to SAP:

    1. This description of the problem (short dump)
    To do this, choose  System -> List -> Save -> Local File (unconverted)
    on the screen you are in now.

    2. A suitable system log
    To do this, call the system log in transaction SM21. Restrict the time
    interval to ten minutes before the short dump and five minutes after
    it. In the display, choose System -> List -> Save -> Local File
    (unconverted).

    3. If these are programs of your own, or modified SAP programs: Source
    code of these programs
    To do this, choose  More Utilities -> Upload/Download -> Download in
    the Editor.

    4. Details regarding the conditions under which the error occurred or
    which actions and input caused the error.

System environment
    SAP Release..... 740
    SAP Basis level 0023

    Application server... <removed>
    Network address...... <removed>
    Operating system... <removed>
    Release.............. <removed>
    Hardware type....... <removed>
    Character length..... 16 Bits
    Pointer length........ 64 Bits
    Work process number... 0
    Shortdump setting. full

    Database server... <removed>
    Database type..... ORACLE
    Database name..... <removed>
    Database user ID SAPSR3

    Terminal.......... <removed>

    Character set C

    SAP kernel....... 753
    Created on....... Jun 11 2020 00:17:53
    Created at....... NT 6.3 9600 x86 MS VC++ 18.00 use-pr20200610
    Database version OCI_11201_SHARE, 11.2.0.3.0
    Patch level....... 602
    Patch text.......

    Database............. ORACLE 11.2.*.*.*, ORACLE 12.*.*.*.*
    SAP database version. 753
    Operating system... Windows NT 6.0, Windows NT 6.1, Windows NT 6.2, Windows NT
     6.3, Windows NT 10.0

    Memory consumption
    Roll.... 0
    EM...... 2004191024
    Heap.... 2050793808
    Page.... 106496
    MM used. 4019973616
    MM free. 22793312

Information on where terminated
    The termination occurred in ABAP program "CL_ABAP_ZIP===================CP",
     in "SAVE". The main program
    was "ZABAPGIT".

    In the source code, the termination point is in line 60 of (Include)
    program "CL_ABAP_ZIP===================CM006".

Source Code Extract

Line  SourceCde

   30     start_offset = XSTRLEN( zip ).
   31
   32     msdos_time = msdos=>to_time( <file>-time ).
   33     msdos_date = msdos=>to_date( <file>-date ).
   34
   35 *   zip data stream
   36     writeX4  zip '504B0304'.                  " local file header signature
   37     write2   zip <ext>-min_extract_version.   " version needed to extract = 2.0 - File is co
   38     write2   zip <ext>-gen_flags.             " general purpose bit flag
   39     write2   zip <ext>-compressed.            " compression method: deflated
   40     write2   zip msdos_time.                  " last mod file time
   41     write2   zip msdos_date.                  " last mod file date
   42     write4   zip <ext>-crc32.                 " crc-32
   43     write4   zip <ext>-compsize.              " compressed size
   44     write4   zip <file>-size.                 " uncompressed size
   45     write2   zip <ext>-filename_len.          " file name length
   46     write2   zip <ext>-extra_len.             " extra field length
   47
   48     CONCATENATE zip <ext>-filename <ext>-extra <ext>-content INTO zip IN BYTE MODE.
   49
   50 *   central directory stream (which has a lare duplicate sequence of zip header)
   51     DATA: dup_offset TYPE i. dup_offset = start_offset + 4.
   52     writeX4  dir '504B0102'.                  " central file header signature
   53     write2   dir 19.                          " version made by (== pkzip 2.04g)
   54     CONCATENATE dir zip+dup_offset(26) INTO dir IN BYTE MODE.  " part which matches exactly
   55     write2   dir  0.                          " file comment length
   56     write2   dir  0.                          " disk number start
   57     write2   dir  0.                          " internal file attributes
   58     write4   dir  0.                          " external file attributes
   59     write4   dir  start_offset.               " relative offset of local header
>>>>>     CONCATENATE dir <ext>-filename  <ext>-extra INTO dir IN BYTE MODE.  " file name + extra
   61
   62   ENDLOOP.
   63
   64 * Write Central Directory
   65   DATA: lines_files TYPE i.    lines_files = LINES( files ).
   66   DATA: xstrlen_dir TYPE i.    xstrlen_dir = XSTRLEN( dir ).
   67   DATA: offset_dir  TYPE i.    offset_dir  = XSTRLEN( zip ).
   68
   69   CONCATENATE zip dir INTO zip IN BYTE MODE.
   70   writeX4  zip '504B0506'.                    " End of central directory
   71   write2   zip  0.                            " number of this disk
   72   write2   zip  0.                            " number of the disk with the start of the cen
   73   write2   zip  lines_files.                  " total number of entries in the central direc
   74   write2   zip  lines_files.                  " total number of entries in the central direc
   75   write4   zip  xstrlen_dir.                  " size of the central directory
   76   write4   zip  offset_dir.                   " offset of start of central directory
   77   write2   zip  0.                            " ZIP file comment length
   78
   79 ENDMETHOD.

Contents of system fields

Name     Val.

SY-SUBRC 0
SY-INDEX 0
SY-TABIX 111649
SY-DBCNT 0
SY-FDPOS 1
SY-LSIND 0
SY-PAGNO 0
SY-LINNO 1
SY-COLNO 1
SY-PFKEY %_CS
SY-UCOMM &ONT
SY-TITLE abapGit
SY-MSGTY
SY-MSGID
SY-MSGNO 000
SY-MSGV1
SY-MSGV2
SY-MSGV3
SY-MSGV4
SY-MODNO 0
SY-DATUM 20210514
SY-UZEIT 115730
SY-XPROG SAPMSSY1
SY-XFORM XAB_WFLUSH

Active Calls/Events

No.   Ty.          Program                             Include                             Line
      Name

   18 METHOD       CL_ABAP_ZIP===================CP    CL_ABAP_ZIP===================CM006    60
      CL_ABAP_ZIP=>SAVE
   17 METHOD       ZCL_ABAPGIT_ZIP===============CP    ZCL_ABAPGIT_ZIP===============CM001    17
      ZCL_ABAPGIT_ZIP=>ENCODE_FILES
   16 METHOD       ZCL_ABAPGIT_ZIP===============CP    ZCL_ABAPGIT_ZIP===============CM002    35
      ZCL_ABAPGIT_ZIP=>EXPORT
   15 METHOD       ZCL_ABAPGIT_ZIP===============CP    ZCL_ABAPGIT_ZIP===============CM004    26
      ZCL_ABAPGIT_ZIP=>EXPORT_PACKAGE
   14 METHOD       ZCL_ABAPGIT_GUI_ROUTER========CP    ZCL_ABAPGIT_GUI_ROUTER========CM00L    46
      ZCL_ABAPGIT_GUI_ROUTER=>ZIP_SERVICES
   13 METHOD       ZCL_ABAPGIT_GUI_ROUTER========CP    ZCL_ABAPGIT_GUI_ROUTER========CM00K    12
      ZCL_ABAPGIT_GUI_ROUTER=>ZIF_ABAPGIT_GUI_EVENT_HANDLER~ON_EVENT
   12 METHOD       ZCL_ABAPGIT_GUI===============CP    ZCL_ABAPGIT_GUI===============CM008    18
      ZCL_ABAPGIT_GUI=>HANDLE_ACTION
   11 METHOD       ZCL_ABAPGIT_GUI===============CP    ZCL_ABAPGIT_GUI===============CM00A     3
      ZCL_ABAPGIT_GUI=>ON_EVENT
   10 METHOD       ZCL_ABAPGIT_HTML_VIEWER_GUI===CP    ZCL_ABAPGIT_HTML_VIEWER_GUI===CM002     3
      ZCL_ABAPGIT_HTML_VIEWER_GUI=>ON_EVENT
    9 METHOD       CL_GUI_HTML_VIEWER============CP    CL_GUI_HTML_VIEWER============CM00K    17
      CL_GUI_HTML_VIEWER=>DISPATCH
    8 METHOD       CL_GUI_CFW====================CP    CL_GUI_CFW====================CM001   138
      CL_GUI_CFW=>DISPATCH
    7 FORM         SAPFGUICNTL_CFW                     SAPFGUICNTL_CFW                       175
      CONTROL_END
    6 FORM         SAPFGUICNTL                         SAPFGUICNTL                           677
      CONTROL_END
    5 FORM         SAPMSSYD                            SAPMSSYD                              481
      %_CTL_END
    4 MODULE (PAI) ZABAPGIT                            <SYSINI>                               22
      %_CTL_END
    3 FORM         ZABAPGIT                            ZABAPGIT_FORMS                         29
      OPEN_GUI
    2 FORM         ZABAPGIT                            ZABAPGIT_FORMS                         14
      RUN
    1 EVENT        ZABAPGIT                            ZABAPGIT                               57
      START-OF-SELECTION

Chosen variables

Name
    Val.

No.      17 Ty.          METHOD
Name  ZCL_ABAPGIT_ZIP=>ENCODE_FILES
IT_FILES
    Table IT_22680[175877x256]
    \CLASS=ZCL_ABAPGIT_ZIP\METHOD=EXPORT\DATA=LT_ZIP
    Table reference: 1580
    TABH+ 0(20) = 60B5844A010000000000000000000000F02F7600
    TABH+20(20) = DF070000000000000000000005AF020044040000
    TABH+40(20) = 0000000000000000000000002C06000098580000
    TABH+60(20) = 05AF020000010000FFFFFFFF04600100F0910000
    TABH+80( 8) = 0200000041060100
    store                   = 0x60B5844A01000000
    statScndKeyAdmin        = 0x0000000000000000
    ext1                    = 0xF02F7600DF070000
    ----- Primary Key -----
    idxPtr                  = 0x0000000000000000
      idxKind               = 0     (ItIndexNone)
      accKind               = 1     (ItAccStandard)
      uniKind               = 2     (ItUniNo)
      keyKind (from pbag)   = 1     (default)
      cmpMode               = 8     (cmpManyEq)
    shmId                   = 0     (0x00000000)
    id                      = 1580  (0x2C060000)
    label                   = 22680 (0x98580000)
    fill                    = 175877 (0x05AF0200)
    leng                    = 256   (0x00010000)
    loop                    = -1    (0xFFFFFFFF)
    xtyp                    = PROG#000352/TYPE#000011
    occu                    = 2     (0x02000000)
```

### PR DESCRIPTION
**Pull request by abap-weasel**

### CODE REVIEW

File: src/objects/core/zcl_abapgit_serialize.clas.abap [hunk 1]
```diff
@@ -224,7 +225,13 @@ CLASS zcl_abapgit_serialize IMPLEMENTATION.
       iv_language         = io_dot_abapgit->get_main_language( )
       ii_log              = ii_log
       iv_force_sequential = lv_force ).
-    APPEND LINES OF lt_found TO ct_files.
+* transfer one by one and delete from original table as APPEND LINES OF
+* leads to doubling memory usage during execution of statement which
+* could result in SYSTEM_NO_ROLL dump
+    LOOP AT lt_found ASSIGNING <ls_found>.
+      APPEND <ls_found> TO ct_files.
+      DELETE lt_found.
```
Comment (larshp) [inline src/objects/core/zcl_abapgit_serialize.clas.abap:233]: suggest making the INDEX explicit in this statement

Comment (abap-weasel) [reply src/objects/core/zcl_abapgit_serialize.clas.abap]: I was actually trained in WDF to do DELETE in Loop without any addition if the current line should be delete. It was expected to be the most robust (no need to check if index is fetched at correct location, no struggling with index vs. tabix) and and fastest ("itab knows its current line"). 

As I struggled to find any evidence on SAP help, I tested this out... and the reality is that DELETE ... INDEX is actually always about 20 % faster (tested on NW 7.00 ... up to S4HANA 2020). 

Will have to look through some own code now.... 😥

Comment (larshp) [reply src/objects/core/zcl_abapgit_serialize.clas.abap]: yea... I think its better to always specify the index explicitly

Before this commit, I was not sure what would happen, does APPEND set the implicit index, or is it taken from the LOOP? Anyhow, now the developer can read what happens without knowing ABAP internal mechanisms

Comment (abap-weasel) [reply src/objects/core/zcl_abapgit_serialize.clas.abap]: For sake of completeness, I just found the proper documentation:

https://help.sap.com/doc/abapdocu_750_index_htm/7.50/en-US/abapdelete_itab_line.htm#!ABAP_ONE_ADD@2@
"This variant is only possible within a LOOP across the same internal table. The current table row of the LOOP is then deleted implicitly."

As per my best understanding (and what I have learnt in WDF), it always deletes the record that is currently processed by LOOP. It does not rely on any system field (index vs. tabix). So it works as expected, but impact on performance still makes me a little sad..

File: src/objects/core/zcl_abapgit_serialize.clas.abap [hunk 2]
```diff
@@ -224,7 +226,14 @@ CLASS zcl_abapgit_serialize IMPLEMENTATION.
...
       iv_force_sequential = lv_force ).
-    APPEND LINES OF lt_found TO ct_files.
+* transfer one by one and delete from original table as APPEND LINES OF
+* leads to doubling memory usage during execution of statement which
+* could result in SYSTEM_NO_ROLL dump
+    LOOP AT lt_found ASSIGNING <ls_found>.
+      lv_index = sy-tabix.
+      APPEND <ls_found> TO ct_files.
+      DELETE lt_found INDEX lv_index.
+    ENDLOOP.
```
Comment (jan-jezek) [inline src/objects/core/zcl_abapgit_serialize.clas.abap:236]: Removing row from table doesn't actually help with the memory.
The allocated memory for table lt_found stays the same even after you delete the row from the table.

If you really want to delete the row and save some memory you would have to exit the loop (maybe after you have deleted 1000 rows), copy the table to new variable (which will allocate new memory space, but only for the remaining rows), free the old table and then continue with the loop.

Using delete without this is pointless and provides just an illusion of better memory use.

Comment (jan-jezek) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:236]: To correct myself:
Deleting rows from the table has effect on memory, only after the table has been fully deleted.
I have run a simple test. Create a stringtab and fill it with numbers from 1 to 100000.
You can check the memory usage using cl_abap_memory_utilities=>get_memory_size_of_object( ).
I delete the rows from the table using index.

Here are the results:
Full table
Half deleted table
Last record still in table
Empty table

As you can see the sizeof_used gradually drops as expected.
But sizeof_alloc is near the end greater by more than a half.
It drops to almost nothing after the last row is deleted, but you can do a free after the loop with same effect.

File: src/objects/core/zcl_abapgit_serialize.clas.abap [hunk 3]
```diff
@@ -224,7 +226,14 @@ CLASS zcl_abapgit_serialize IMPLEMENTATION.
...
       ii_log              = ii_log
       iv_force_sequential = lv_force ).
-    APPEND LINES OF lt_found TO ct_files.
+* transfer one by one and delete from original table as APPEND LINES OF
+* leads to doubling memory usage during execution of statement which
+* could result in SYSTEM_NO_ROLL dump
+    LOOP AT lt_found ASSIGNING <ls_found>.
+      lv_index = sy-tabix.
+      APPEND <ls_found> TO ct_files.
+      DELETE lt_found INDEX lv_index.
```
Comment (jan-jezek) [inline src/objects/core/zcl_abapgit_serialize.clas.abap:235]: Storing and using the index to delete from table is pointless as the sy-tabix will always be 1 if you are deleting from the table.

File: src/objects/core/zcl_abapgit_serialize.clas.abap (hunk 2 above)
Comment (larshp) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:236]: in the debugger, try triggering the garbage collector manually?
image

Comment (jan-jezek) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:236]: I tried manually in the debugger and using cl_abap_memory_utilities=>do_garbage_collection( ).
None seems to have an effect, allocated memory stays still same.

Comment (jan-jezek) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:236]: And just one more correction from my side...
Exiting loop at 50000 rows deleted and creating new variable (using DATA(new_table) = old_table.) allocates the same memory as was for old_table (the bigger number) which is not what we want.
So my workaround is not working.

However what works is exporting the table to shared memory, freeing it and importing it back (done after 50000 rows were deleted).
Before export:
before_export
After import:
after_import

Comment (abap-weasel) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:236]: Well, having no deeper knowledge on memory internals, I'd like to show you what I observed during testing of the change in SAPGUI debugger:

grafik
-> "Dyn. Memory Object (Used)" and "Total" got quite lower. Based on that I commited the change.

After your the comments I am not sure what these numbers actually mean.... Shall I change it back to to APPEND LINES OF?

Comment (larshp) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:236]: @abap-weasel is above only for the APPEND LINES OF change? or the full PR scope?

Comment (abap-weasel) [reply src/objects/core/zcl_abapgit_serialize.clas.abap:236]: That is only for the APPEND LINES OF change (with some reduced package size). The Total goes a little bit up during append and immediatly reduced by delete again.

Regarding full PR scope, I did not find a good way to validate that exactly. I tried to get the numbers by doing some dumps to SMI transaction: 

grafik

File: src/objects/core/zcl_abapgit_serialize.clas.abap (hunk 3 above)
Comment (abap-weasel) [reply src/objects/core/zcl_abapgit_serialize.clas.abap]: You're right. Done.

**Comment by mbtools** [comment]
it will be much faster to assign your >100k objects to several packages and export them separately than trying to figure out how to  reduce memory consumption.

File: src/objects/core/zcl_abapgit_serialize.clas.abap (hunk 2 above)
Comment (jan-jezek) [reply src/objects/core/zcl_abapgit_serialize.clas.abap]: IMHO I would remove the delete statement and I would just do the append.

Comment (larshp) [reply src/objects/core/zcl_abapgit_serialize.clas.abap]: yea, I agree on that

for the ZIP scenario it should fit into memory anyhow, as the memory requirement is doubled later

Comment (abap-weasel) [reply src/objects/core/zcl_abapgit_serialize.clas.abap]: sure, then let's get back to original APPEND version

**Review by larshp — APPROVED** [review]
looks good to me, thanks 👍

**Comment by larshp** [comment]
welcome to abapGit as a contributor 🎉

**Pull request merged**